docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1565,53 +1565,85 @@ fits the encoded integer value. This representation achieves three goals:
-restrictions on the length of the string itself. Instead, the server verifies
-that the highest-order non-zero bit position in the string is within the
-bounds established by the P4Info bitwidth. If `p4_bitwidth` is the P4Info-
-specified bitwidth for a P4 object value, then the server computes the number of
-binary string bytes required to represent the maximum integer of `p4_bitwidth`
-bits as:
-~ Begin CPP
-binary_string_bytes = (p4_bitwidth + 7) / 8;
-~ End CPP
-The server's decision to accept or reject the binary string depends on the
-string's content and its actual length in the P4Runtime request, as follows:
-
-* If the string's actual byte length is shorter than `binary_string_bytes`
-  but greater than zero, the server always accepts the string and treats the
-  missing bytes as zeroes in the most-significant positions.
-
-* If the string's actual byte length is longer than `binary_string_bytes`, the
-  server rejects any string where the extra high-order bytes contain non-zero
-  values. Otherwise, the server truncates the extra bytes from the front
-  of the string and processes the remainder of the string as if it contains
-  exactly `binary_string_bytes`, as described below.
-
-* If the string's original or truncated byte length is equal to
-  `binary_string_bytes`, then the server must also consider the P4 integer
-  type's byte alignment. Byte-aligned strings that satisfy the expression:
-~ Begin CPP
-  (binary_string_bytes * 8 == p4_bitwidth)
-~ End CPP
-  are always valid. For non-aligned strings, the server requires
-  the client to fill the high-order pad bits with zeroes, and the server
-  rejects any binary strings with non-zero pad bits.
-
-* If the string's byte length is zero, the server always rejects the string.
+Any additional bits in the bytes sent for an unsigned integer value
+(type `bit<W>`) must be 0.  If additional bytes are transmitted above
+the `encoded_string_size` minimum required, they must be filled with
+0.
+
+Any additional bits in the bytes sent for a signed integer value (type
+`int<W>`) must be copies of the sign bit, i.e. 0 for non-negative
+values, or 1 for negative values. If additional bytes are transmitted
+above the `encoded_string_size` minimum required, they must be filled
+with copies of the sign bit, i.e. 0 for non-negative values, or 0xff
+for negative values. In 2's complement representation, this is called
+"sign extension", and leaves the numeric value represented unchanged.
+
+Upon receiving a binary string, the P4Runtime receiver (whether the
+server or the client) does not impose any restrictions on the length
+of the string itself. Instead, the receiver verifies that the value
+encoded by the string fits within the expected type (signed or
+unsigned) and P4Info-specified bitwidth for the P4 object value.
+
+For received values expected to fit within a `bit<W>` type, the value
+is in range if, after removing all most significant 0 bits, the value
+remaining is `W` bits or less.
```

Comment:
I think it's more accurate to specify the value width here, i.e "the remaining value's width is W bits or less".